### PR TITLE
Fixes #21457 - Allow password based auth

### PR DESCRIPTION
**Pull request by adamruzicka**

### CODE REVIEW

**Comment by iNecas** [comment]
Rubocop is not happy

**Comment by iNecas** [comment]
Also, waiting for future and recurring logic to work with this.

**Comment by adamruzicka** [comment]
Starting with ed37691 this PR requires https://github.com/theforeman/foreman-tasks/pull/297 .

File: app/lib/actions/remote_execution/run_hosts_job.rb [hunk 1]
```diff
@@ -38,6 +41,10 @@ def create_sub_plans
         end
       end
 
+      def finalize
+        input.delete(:secrets)
```
Comment (iNecas) [inline app/lib/actions/remote_execution/run_hosts_job.rb]: Is this still needed, with `HideSecrets` in place?

Comment (adamruzicka) [reply app/lib/actions/remote_execution/run_hosts_job.rb]: The `HideSecrets` middleware only prevents showing the secrets in the webui. Anyone with access to the database could still look there to find the secrets. For this reason we should delete them by hand when they're not needed anymore

Comment (iNecas) [reply app/lib/actions/remote_execution/run_hosts_job.rb]: Oh, I"ve seen https://github.com/theforeman/foreman-tasks/pull/297/files#diff-03dde90ec7e9caa322015dee15f4e956R103 also doing some wiping, but now I see it is run_host_job vs. run_hosts_job. Nm then

**Comment by adamruzicka** [comment]
Update:
- [ ] (imagine this box half checked) successful run ditches the secrets properly: however, if the execution fails they stay in the db, but at least they're encrypted now
- [x] password and passphrase are now stored as attributes of job invocation and are encrypted before hitting the db
- [ ] encrypted parameters: to be done

**Comment by ares** [comment]
Note that http://projects.theforeman.org/issues/3378 mentiones smart variables. I don't think we should further extend parameters. They support hidden values already. IMHO this is good to be shipped without parameters encryption. The encryption does not provide much, the value could be easily decrypted and read.

File: app/models/concerns/foreman_remote_execution/encrypted_value_extensions.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+module ForemanRemoteExecution
```
Comment (iNecas) [inline app/models/concerns/foreman_remote_execution/encrypted_value_extensions.rb]: Any reason we can't use the `Encryptable` module from the Foreman? https://github.com/theforeman/foreman/blob/b5eefd928921ddab7fce2b03882e1b8afa5f37f5/app/models/concerns/encryptable.rb#L2

**Comment by iNecas** [comment]
I believe there is an added value for the encryption, as on can't at least decode the credentials from db dump. Anyway, I would be good with just using the secret params for that, and I don't insist on this being blocker for this PR. However, I what I would like to see before we proceed is:

1. global setting for default credentials
2. the setting overridable by host params, similarly as we have for `remote_execution_ssh_user` and others, for the sake of consistency

**Comment by adamruzicka** [comment]
Added encrypted global setting and possibility to override it with host params

**Comment by ares** [comment]
needs rebasing :-/

**Comment by adamruzicka** [comment]
@ares rebased

File: app/lib/actions/remote_execution/run_host_job.rb [hunk 3]
```diff
@@ -47,10 +48,17 @@ def plan(job_invocation, host, template_invocation, proxy_selector = ::RemoteExe
         raise _('Failed rendering template: %s') % renderer.error_message unless script
 
         provider = template_invocation.template.provider
-        hostname = provider.find_ip_or_hostname(host)
+
+        secrets = { :ssh_password => job_invocation.password || provider.ssh_password(host),
+                    :key_passphrase => job_invocation.key_passphrase || provider.ssh_key_passphrase(host) }
```
Comment (iNecas) [inline app/lib/actions/remote_execution/run_host_job.rb]: This gives me `Failed to initialize: Dynflow::ExecutionPlan::Steps::Error - private method `ssh_password' called for SSHExecutionProvider:Class`

**Comment by iNecas** [comment]
I've released the foreman-tasks 0.11.1 with the required changes. I've tested this (including future and recurring execution) and everything works well. Please bump the version requirement here and rebase, so that we can get the green tests.

**Comment by iNecas** [comment]
Test failure related

**Comment by iNecas** [comment]
Thanks @adamruzicka

**Comment by ares** [comment]
:clap: :+1: :1st_place_medal:

**Pull request merged**